### DISCUSSION TITLE
Data Augmentation Rework - using albumentations

### DISCUSSION THREAD

**Discussion by florenttaralle** [ultralytics/yolov5#7225] [Ideas]
Hi there, 

Yolov5 is both such a great architecture and a great training environment. 
It is so easy to train and test new models from scratch. I love working with it.
But when I have to look at data augmentation en preparation, I feel realy confused. 
The code is realy hard to understand and upgrade. That said, I noticed upgrades there, but not enougth !

As a first exemple, I recently had to add a random bgr_to_gray transform.
As a second exemple, I would like to implemente rotated-bbox prediction. That requires a lot of work but data-augmentation seems to be the hardest to change at first.

I wonder if we could rework the data-augmentation part of Yolov5 using [albumentations](https://albumentations.ai/).
1. using pre-implemented augmentations and linking them to the config yaml file (hyp)
2. allowing to import additionnal albumentation-compatible transforms from python scripts/folders.

To be noted that I noticed that [albumentations](https://albumentations.ai/) is partialy used in [augmentations.py](https://github.com/ultralytics/yolov5/blob/master/utils/augmentations.py#L16) but as a single all-in-one transform, sadly not as a generic data-augmentation framework. 

So I start this discussion to know if other think (1) this is a good idea and if (2) albumentations is the good choice.

**Comment by glenn-jocher** [comment]
@florenttaralle hi, thank you for your feature suggestion on how to improve YOLOv5 🚀! Yes, it might make sense to consolidate the augmentations together more with albumentations, as there is a certain level of duplication currently. Albumentations can not fully replace the YOLOv5 augmentations though as it doesn't offer all types, like Mosaic, and the team does not plan to support those in future that would require multiple images (i.e. Mosaic).


The fastest and easiest way to incorporate your ideas into the official codebase is to submit a Pull Request (PR) implementing your idea, and if applicable providing before and after profiling/inference/training results to help us understand the improvement your feature provides. This allows us to directly see the changes in the code and to understand how they affect workflows and performance.

Please see our ✅ [Contributing Guide](https://github.com/ultralytics/yolov5/blob/master/CONTRIBUTING.md) to get started.

**Reply by florenttaralle** [reply]
> Albumentations can not fully replace the YOLOv5 augmentations though as it doesn't offer all types, like Mosaic, and the team does not plan to support those in future that would require multiple images (i.e. Mosaic).

Yes `Mosaic` is specific. In my opinion it is not a **transform** per say. Thus it could never legitimately be integrated in albumentations' framework. It is a more like a pre and/or post processing that aggregates items together.

I am considering implementing what i proposed here. 
But i prefer hasking for other's feedbacks first to avoid working on a useless or ill-defined proposition :)

**Reply by florenttaralle** [reply]
@glenn-jocher 
> The fastest and easiest way to incorporate your ideas into the official codebase is to submit a Pull Request (PR) implementing your idea, and if applicable providing before and after profiling/inference/training results to help us understand the improvement your feature provides. This allows us to directly see the changes in the code and to understand how they affect workflows and performance.
> 
> Please see our white_check_mark [Contributing Guide](https://github.com/ultralytics/yolov5/blob/master/CONTRIBUTING.md) to get started.

Hi, 
I started working on the code. 
I tried to make a new branch from the master (git checkout -b) then to push my work-in-progress on it. 
But I got an authentification error (403). I used a PAC (personal access token) as required by github.
Am-I doing this wrong ?
